otherlibs/win32unix/createprocess.c

```diff
@@ -24,33 +24,45 @@
 
 static int win_has_console(void);
 
+static int ensure_inheritable(HANDLE h, HANDLE * hStd, HANDLE * to_close)
+{
+  DWORD flags;
+  HANDLE hp;
+
+  if (! GetHandleInformation(h, &flags))
+    return 0;
+  hp = GetCurrentProcess();
+  if (! (flags & HANDLE_FLAG_INHERIT)) {
+    if (! DuplicateHandle(hp, h, hp, hStd, 0, TRUE, DUPLICATE_SAME_ACCESS))
+      return 0;
+    to_close = hStd;
+  } else {
+    *hStd = h;
+  }
+  return 1;
+}
+
 static DWORD do_create_process_native(wchar_t * exefile, wchar_t * cmdline,
-                                      wchar_t * env, HANDLE fd1, HANDLE fd2,
-                                      HANDLE fd3, HANDLE * hProcess)
+                                      wchar_t * env, HANDLE fd0, HANDLE fd1,
+                                      HANDLE fd2, HANDLE * hProcess)
 {
   PROCESS_INFORMATION pi;
   STARTUPINFO si;
   DWORD flags, err;
-  HANDLE hp;
+  HANDLE to_close[3] =
+      {INVALID_HANDLE_VALUE, INVALID_HANDLE_VALUE, INVALID_HANDLE_VALUE};
```

Comment:
I wouldn't use an array here, it obfuscates matters. Let just call them `to_close1`, `to_close2`, `to_close3` (and we can revert the above renaming `fd1 -> fd0`, etc).